### DISCUSSION TITLE
alchemyProvider Security - Ideas?

### DISCUSSION THREAD

**Discussion by a-znk** [wevm/wagmi#1345] [Q&A]
I'm currently working on a project that's using the alchemyProvider, and I'd like to be following best security practices for my API key.

At the moment, the project is built using NextJS and has been deployed on Vercel. The Alchemy dashboard of course provides the option to create an Allowlist. However, Vercel IP addresses are dynamic, so I can't use their recommended method of allowlisting the site IP address, and allowlisting the domain is not reliable.

I was thinking of building some kind of an API route using NextJS's API routes. However, I wanted to open it up to see if anybody had a more effective way?

Any insight would be appreciated!

**Comment by rsodre** [comment]
I also had a lot of issues with Vercel and Alchemy domain allowlist.

Did you try adding your contracts to the **addresses allowlist**?

**Reply by a-znk** [reply]
I haven't tried this approach yet. Would this have the same result as IP/Domain allowlisting (i.e. only my DAPP with my approved contract will work with my key), or are there other loopholes for people to use the key? (understanding there are practically ALWAYS some loopholes in security...)
